### ISSUE TITLE
POST requests in shinyproxy

### ISSUE THREAD

**Issue by shosaco** [openanalytics/shinyproxy#184]
Dear community,

I am trying to use Shinyproxy with a small Flask app and upload a file via a form/POST request in my Flask app works fine when the app is executed directly but fails when Shinyproxy hosts the Flask app. I tracked the issue down to the point that Shinyproxy is not executing the form POST request properly:

**app.py:**

    from flask import Flask, request, render_template
    
    app = Flask(__name__, static_url_path="/static")
    
    @app.route('/', methods=['GET', 'POST'])
    def index():
        if request.method == 'GET':
            return render_template('index.html')
        else:
            return "POST request received"
    
    app.run(host='0.0.0.0', port=3838)

**index.html:**

    <html>
    <body>
        <form method=post enctype=multipart/form-data>
            <input type="file" name="file"/>
            <input type = "submit" value="Upload">  
        </form>  
    </body>
    </html>

returns, after clicking on "Upload", what one would expect, the GET and the POST request:

    10.81.71.42 - - [23/Dec/2019 14:37:28] "GET / HTTP/1.1" 200 -
    10.81.71.42 - - [23/Dec/2019 14:37:30] "POST / HTTP/1.1" 200 -

Running the exact same app in Shinyproxy via the Dockerfile

    FROM python:3
    RUN pip install flask werkzeug
    RUN mkdir /templates
    COPY ["index.html", "/templates"]
    COPY app.py /
    EXPOSE 3838
    CMD ["python", "app.py"]

returns the same but without the POST line:

    172.17.0.1 - - [23/Dec/2019 14:39:42] "GET / HTTP/1.1" 200 -

The **shinyproxy.log** says

    2019-12-23 14:42:33.682 DEBUG 17832 --- [XNIO-2 I/O-1] io.undertow.server.handlers.proxy        : Sent request ClientRequest{path='/', method=POST, protocol=HTTP/1.1} to target 10.81.71.42 for exchange HttpServerExchange{ POST /proxy_endp
    oint/b933863e-9fad-4d00-a657-034ede313e34/ request {Accept=[text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9], Accept-Language=[de-DE,de;q=0.9,en-US;q=0.8,en;q=0.
    7], Cache-Control=[max-age=0], Accept-Encoding=[gzip, deflate], Origin=[http://192.168.76.81:8080], User-Agent=[Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36], Connectio
    n=[keep-alive], Content-Length=[188], Content-Type=[multipart/form-data; boundary=----WebKitFormBoundarymPv6WtCTpGZQqRbF], Cookie=[JSESSIONID=CZxjGBM7BW597wysbBABgNWZL2x7qvsfujWVERgR], Referer=[http://192.168.76.81:8080/app_direct/flask_t
    est/], Upgrade-Insecure-Requests=[1], Host=[192.168.76.81:8080]} response {Expires=[0], Cache-Control=[no-cache, no-store, max-age=0, must-revalidate], X-XSS-Protection=[1; mode=block], X-Content-Type-Options=[nosniff], Pragma=[no-cache]}
    }                                                                                                                                                                                                                                             
    2019-12-23 14:42:33.683 DEBUG 17832 --- [XNIO-2 I/O-1] io.undertow.request.io                   : Fixed length stream closed with with 188 bytes remaining                                                                                    
    2019-12-23 14:42:33.683 DEBUG 17832 --- [XNIO-2 I/O-1] i.u.client.http.HttpClientExchange       : request terminated for request to localhost/127.0.0.1:20000 /                                                                               
    2019-12-23 14:42:33.683 ERROR 17832 --- [XNIO-2 I/O-1] io.undertow.proxy                        : UT005028: Proxy request to /proxy_endpoint/b933863e-9fad-4d00-a657-034ede313e34/ failed                                                     
                                                                                                                                                                                                                                                  
    io.undertow.server.TruncatedResponseException: null                                             

I know the use case of Shinyproxy might not be to host Flask apps, however, I love Shiny and Shinyproxy and thought this might be a good use case.

Best regards
shosaco

**Comment by vpanfilov** [comment]
It may be the same as bugs https://github.com/openanalytics/shinyproxy/issues/140 and https://github.com/openanalytics/shinyproxy/issues/170.

Try to revert to Shiny Proxy 2.1.0.

**Comment by tverbeke** [comment]
> I know the use case of Shinyproxy might not be to host Flask apps, however, I love Shiny and Shinyproxy and thought this might be a good use case.

It's definitely part of ShinyProxy's use case - no doubt about that!

**Comment by shosaco** [comment]
> It may be the same as bugs #140 and #170.
> 
> Try to revert to Shiny Proxy 2.1.0.

That solved the problem. Will there be a fix in a newer version available?

Thanks!

**Comment by sound118** [comment]
> > I know the use case of Shinyproxy might not be to host Flask apps, however, I love Shiny and Shinyproxy and thought this might be a good use case.
> 
> It's definitely part of ShinyProxy's use case - no doubt about that!

Is there a demo/tutorial showing how to deploy pure flask app with shinyproxy ?

**Comment by shosaco** [comment]
> Is there a demo/tutorial showing how to deploy pure flask app with shinyproxy ?

My question was such a minimal example. Only thing missing is the application.yaml, which should state the same as the docker CMD command (`container-cmd: ["python", "app.py"]`)

**Comment by kunal15145** [comment]
this error is present only in case of multipart/form-data not in the upload using other content type. Any update would be appreciated without deprecating to 2.1.0 version

**Comment by kunal15145** [comment]
I know this might be a hack though but for small files this is working, just figured out. Add this to your application.yml
`spring:
    servlet:
        multipart:
            enabled: false`
@shosaco

**Comment by shosaco** [comment]
Thanks, that indeed works for small files. For large files, I need [this entry](https://www.shinyproxy.io/configuration/#uploading-large-files), which excludes your setting :/

**Comment by kunal15145** [comment]
I don't think that works. As long as the multipart is enabled, we will get the truncated exception.

**Comment by shosaco** [comment]
That's what I'm saying. My link points to a setting that specifies the file size of multipart. Your solution disables multipart. These two settings are exclusive, so for large files there is still not solution.

**Comment by vpanfilov** [comment]
See https://github.com/openanalytics/shinyproxy/issues/140#issuecomment-633964720. It seems that this fix would be able to fix this issue. I am waiting for the new release.

**Comment by shosaco** [comment]
I was hoping the fix for #140 (2.3.1) would also fix this issue, but unfortunately it doesn't.

**Comment by tverbeke** [comment]
HI @shosaco, Thanks for the confirmation - on our plate now. I have no knowledge of flask, but is this relevant:

https://github.com/openanalytics/shinyproxy-dash-demo/blob/master/app/app.py#L46

?

**Comment by shosaco** [comment]
> HI @shosaco, Thanks for the confirmation - on our plate now. I have no knowledge of flask, but is this relevant:
> 
> https://github.com/openanalytics/shinyproxy-dash-demo/blob/master/app/app.py#L46
> 
> ?

Dear tverbeke, thanks for the update. It does not solve the issue. For convenience, I created a gist with the relevant test files:

[https://gist.github.com/shosaco/a242161e7c4ed7534f4df128906472af](https://gist.github.com/shosaco/a242161e7c4ed7534f4df128906472af)

**Comment by LEDfan** [comment]
Hi @shosaco 

Thank you for you extensive report and investigation. 
This should now be fixed in the latest release [ShinyProxy 2.4.0](https://shinyproxy.io/downloads/#240) .
A few remarks:
 - we configured Spring so that it no longer touches multi-part requests. This corresponds to the behavior as in ShinyProxy 2.1.0
 - because spring no longer touches multi-parts requests it is no longer necessary to configure the max upload size using:
```
spring:
  servlet:
    multipart:
      max-file-size: 200MB
      max-request-size: 200MB
```
 - we extensively tested this using the upload functionality of [rstudio](https://github.com/openanalytics/shinyproxy-rstudio-ide-demo). Rstudio is also using multi-part of uploads. We were able to upload big files of at least 1GB without any problems
 - it is important that any reverse proxy or loadbalancer in front of ShinyProxy is correctly configured to handle large file uploads
 - the example Flask app you provided isn't able to handle large uploads (or at least not in our tests). However, as I mentioned before, rstudio is perfectly able to handle it with the latest changes.\

Feel free to open a new issue when you encounter an issue.
